### PR TITLE
Change the IRC interface for easier usage and channel selection

### PR DESCRIPTION
**Pull request by tilpner**
This PR adds support to select channels for the IRC bot, without introducing
a huge amount of additional bots.

The bot can now be configured to listen for certain patterns and to use
a certain channel to evaluate the code.

My current irc.yaml looks like:

```
- nickname: "playbot-dev"
  server: irc.mozilla.org
  port: 6667
  channels:
    - "#rust"
  keys: [null, "hunter2"]
  password: abc123abc
  triggers:
    #- [ "playbot:(.*)", "stable", True ]
    #- [ "rusti:(.*)", "stable", True ]
    #- [ "playbot-mini:(.*)", "stable", False ]
    #- [ "rusti-mini:(.*)", "stable", False ]
    - [ ">>(.*)", "stable", True ]
    - [ "n\\s*>>(.*)", "nightly", True ]
    - [ "nightly\\s*>>(.*)", "nightly", True ]
    - [ "b\\s*>>(.*)", "beta", True ]
    - [ "beta\\s*>>(.*)", "beta", True ]
    - [ "s\\s*>>(.*)", "stable", True ]
    - [ "stable\\s*>>(.*)", "stable", True ]
```

The commented triggers can be uncommented in order to continue listening
like it did before this change.

The only drawback I could find, is the loss of tab-completion for the
`playbot-mini` nickname, but that shouldn't be much of a problem, considering
this was the case for the rusti[-mini] names, and there now are new, shorter, ways
to invoke playbot.

I recommend the addition of short triggers for the "playbot-mini" mode.

### CODE REVIEW

**Comment by alexcrichton** [comment]
This looks like it's also adding some form of feature detection via the normal repo's feature detection script, could that be split off into a separate PR?

**Comment by tilpner** [comment]
@alexcrichton If I was to *split off into a separate PR*, would I have to have a second branch and commit to that one, then PR that new branch?

**Comment by alexcrichton** [comment]
Yeah you'll need to remove the commit from this branch and make a new branch with that commit on it to split it off. You can then force-push the branch of this PR to update the PR.

**Comment by tilpner** [comment]
@alexcrichton I'm sorry, I tried and failed several times. It would be easier if the commit in question was the last commit, but it isn't. Why is that needed?

**Comment by alexcrichton** [comment]
Right now I don't think we don't want to copy scripts from the main distribution and slightly modify them, as I believe there are known deficiencies for feature detection. Additionally it's not clear to me that the playpen should be dealing with features just yet. I'd prefer to consider all that independently of minor configuration changes.

**Comment by tilpner** [comment]
@alexcrichton I deleted the feature detection stuff. Check again?

File: bot.py [hunk 1]
```diff
@@ -41,15 +43,49 @@ def pastebin(command):
...
+    r = requests.post("https://api.github.com/gists",
+            json = {
+                "description": "Shared via Rust Playground",
+                "public": True,
+                "files": {
+                    "playbot.rs": {
+                        "content": code
+                    }
+                }
+            })
```
Comment (alexcrichton) [inline bot.py]: I believe the rate limits for creating gists are quite restrictive, so I believe we're going to need an oauth token here in one form or another to increase our rate limit.

File: bot.py [hunk 2]
```diff
@@ -41,15 +43,49 @@ def pastebin(command):
...
         return "failed to shorten url"
 
-def evaluate(code, nickname):
-    if nickname == "rusti" or nickname == "playbot":
-        version, _ = playpen.execute("stable", "/bin/dash",
+def pastebin(channel, code):
+    if channel == "nightly":
+        return gist(channel, code)
+    else:
+        return bitly(code)
```
Comment (alexcrichton) [inline bot.py]: Why does the method of creating a link depend on the channel?

File: bot.py [hunk 3]
```diff
@@ -60,26 +96,27 @@ def evaluate(code, nickname):
 
     for line in lines:
         if len(line) > 150:
-            return pastebin(code)
+            pastebin(channel, code)
```
Comment (alexcrichton) [inline bot.py]: How come the `return` was removed here?

File: bot.py [hunk 4]
```diff
@@ -60,26 +96,27 @@ def evaluate(code, nickname):
...
 
 class RustEvalbot(irc.client.SimpleIRCClient):
-    def __init__(self, nickname, channels, keys):
+    def __init__(self, nickname, channels, keys, triggers):
         irc.client.SimpleIRCClient.__init__(self)
         irc.client.ServerConnection.buffer_class = irc.buffer.LenientDecodingLineBuffer
         self.nickname = nickname
         self.channels = channels
         self.keys = keys
+        self.triggers = [(re.compile(r), c, wt) for (r, c, wt) in triggers]
```
Comment (alexcrichton) [inline bot.py]: Can this be in a dictionary instead of a list of triples?

**Comment by alexcrichton** [comment]
Thanks @tilpner!

File: bot.py (hunk 2 above)
Comment (tilpner) [inline bot.py]: Oh, right, that's an artifact of the feature detection.
bit.ly will not accept too long URLs, and the feature declaration alone was too long for bit.ly,
thus required a gist.

Shall I leave it in, or default to one of the two?

File: bot.py (hunk 1 above)
Comment (tilpner) [inline bot.py]: [This would indeed be an issue](https://developer.github.com/v3/#rate-limiting).
5000 gists per hour should be fine for the IRC channel though, unless someone abuses it...

File: bot.py (hunk 3 above)
Comment (tilpner) [inline bot.py]: I'm not sure, should be fixed now.

File: bot.py (hunk 4 above)
Comment (tilpner) [inline bot.py]: I guess it could be, but I'm not a fan of arbritary string keys. 
Also, the configuration the irc.yaml might be shorter like this (though shorter != better).

Is it necessary?

File: bot.py (hunk 1 above)
Comment (tilpner) [inline bot.py]: I think I added it, and it seems to work. I didn't stress-test it though.

File: bot.py (hunk 2 above)
Comment (alexcrichton) [inline bot.py]: For now let's stick to just bitly links and leave gisting for features to a separate PR

File: bot.py (hunk 4 above)
Comment (alexcrichton) [inline bot.py]: I don't like the implicit meaning of each element of the tuple where the only definition is the variable names here. Naming the keys should make their purpose much clearer.

File: bot.py [hunk 5]
```diff
@@ -28,28 +29,64 @@
...
+def bitly(command):
     bitly = "https://api-ssl.bitly.com/v3/shorten"
     server = "http://play.rust-lang.org/?"
 
     params = urlencode({"code": command, "run": 1})
     url = server + params
+
     r = requests.get(bitly,
-                     params={"access_token": shorten_key.key, "longUrl": url})
+                     params={"access_token": shorten_key.bitly, "longUrl": url})
```
Comment (alexcrichton) [inline bot.py]: Can you update the README to reflect this change?

**Comment by crumblingstatue** [comment]
@tilpner Are you still interested in finishing this? I am interested in this feature. I can try and finish it if you aren't anymore.

**Comment by tilpner** [comment]
@crumblingstatue Yup, definitely am. For some reason, I forgot to finish this though, so thanks for reminding me. :)

**Comment by tilpner** [comment]
@alexcrichton I updated for the password changes. Have another look now?

**Comment by alexcrichton** [comment]
Oh dear this took me way longer than it should to have gotten to this, thanks @tilpner!

**Comment by alexcrichton** [comment]
Er actually, sorry I had to revert this in #156 because there's some changes that I don't think were brought over. It looks like this removes the ability from one config file to start up multiple connections (e.g. this is just one nick/server combination) which we currently use to log in as playbot and playbot-mini on both mozilla IRC as well as freenode.

Also, could you let me know what the configuration is to migrate over from an old one? Our current configuration looks like:

```yml
- server: irc.mozilla.org
  port: 6667
  channels:
    - "#rust"
    - "#rust-fr"
    - "#rust-gamedev"
    - "#rust-internals"
    - "#rust-offtopic"
    - "#rust-osdev"
    - "#rust-webdev"
    - "#servo"
    - "#rustc"
    - "#rust-libs"
    - "#rust-lang"
    - "#cargo"
    - "#rust-de"
    - "#rust-beginners"
  keys: [null, null, null, null, null, null, null, null, null, null, null, null, null, null]
  password: password
- server: irc.freenode.net
  port: 6667
  channels:
    - "#rust"
  keys: [null]
  password: password
```

**Comment by tilpner** [comment]
Yes, that is correct. This won't start multiple bots at once, but instead allow to listen for multiple configurable triggers.

This PR is so old, I'm unfortunately unable to test if this still works, because I'm not using Arch anymore.

Here's the adapated irc.yaml from a backup of this project.

```yaml
nickname: "playbot"
server: irc.mozilla.org
port: 6667
channels:
    - "#rust"
    - "#rust-fr"
    - "#rust-gamedev"
    - "#rust-internals"
    - "#rust-offtopic"
    - "#rust-osdev"
    - "#rust-webdev"
    - "#servo"
    - "#rustc"
    - "#rust-libs"
    - "#rust-lang"
    - "#cargo"
    - "#rust-de"
    - "#rust-beginners"
keys: [null, null, null, null, null, null, null, null, null, null, null, null, null, null]
password: password
templates:
  - &template "
        #![allow(dead_code, unused_variables)]

        static VERSION: &'static str = \"%(version)s\";

        fn show<T: std::fmt::Debug>(e: T) { println!(\"{:?}\", e) }

        fn main() {
            show({
                %(input)s
            });
        }"
  - &no_template ""
default_template: *template
triggers:
  - template: *template
    channel: "stable"
    triggers:
        - "playbot:(.*)"
        - "rusti:(.*)"
        - ">>(.*)"
        - "s\\s*>>(.*)"
        - "stable\\s*>>(.*)"

  - template: *template
    channel: "beta"
    triggers:
        - "b\\s*>>(.*)"
        - "beta\\s*>>(.*)"

  - template: *template
    channel: "nightly"
    triggers:
        - "n\\s*>>(.*)"
        - "nightly\\s*>>(.*)"

  - template: *no_template
    channel: "stable"
    triggers:
        - "playbot-mini:(.*)"
        - "rusti-mini:(.*)"
        - ">(.*)"
        - "s\\s*>(.*)"
        - "stable\\s*>(.*)"

  - template: *no_template
    channel: "beta"
    triggers:
        - "b\\s*>(.*)"
        - "beta\\s*>(.*)"

  - template: *no_template
    channel: "nightly"
    triggers:
        - "n\\s*>(.*)"
        - "nightly\\s*>(.*)"
```

This setup will keep playbot and playbot-mini on stable, and should not break most peoples workflow,
including private messages. It will, however, break tab-completion for playbot-mini, but that should be an acceptable tradeoff for the channel selection feature.

**Comment by alexcrichton** [comment]
Yeah I'm a little worried with the playbot-mini tab completion, and we also have playbot connected to freenode right now, so this would unfortunately break that as well :(

**Comment by tilpner** [comment]
I've searched through 230M of IRC logs (Mozilla only), and found 10676 invocations of playbot, and only 168 invocations of playbot-mini.

People will still be able to use playbot-mini with the same functionality, even with the same trigger if they wish, their clients just won't tab-complete the -mini anymore.

This isn't new, people have been typing "rusti: " on their own, without tab-completion.

Given the rather infrequent usage of playbot-mini, is this really a problem?

**Comment by alexcrichton** [comment]
I'd like to keep playbot-mini for now, and there's also the issue of how we're currently connected to both mozilla IRC as well as freenode IRC.
